Sources/SwiftParserDiagnostics/ParseDiagnosticsGenerator.swift

```diff
@@ -171,26 +171,38 @@ public class ParseDiagnosticsGenerator: SyntaxAnyVisitor {
     return handledNodes.contains(node.id)
   }
 
-  /// Utility function to emit a diagnostic that removes a misplaced token and instead inserts an equivalent token at the corrected location.
+  /// Utility function to emit a diagnostic that removes a misplaced node and instead inserts an equivalent node at the corrected location.
   ///
-  /// If `incorrectContainer` contains some tokens that satisfy `unexpectedTokenCondition`, emit a diagnostic with message `message` that marks this token as misplaced.
-  /// If `correctTokens` contains missing tokens, also emit a Fix-It with message `fixIt` that marks the unexpected token as missing and instead inserts `correctTokens`.
-  public func exchangeTokens(
+  /// If `incorrectContainer` contains some nodes that satisfy `unexpectedNodeCondition`, emit a diagnostic with
+  /// message `message` that marks the present tokens in this node as misplaced.
+  ///
+  /// If `correctNodes` contains missing tokens, also emit a Fix-It with message `fixIt` that marks the present tokens
+  /// in unexpected nodes as missing while inserting the missing tokens in `correctNodes`.
+  public func exchangeNodes<S: SyntaxProtocol>(
     unexpected: UnexpectedNodesSyntax?,
-    unexpectedTokenCondition: (TokenSyntax) -> Bool,
-    correctTokens: [TokenSyntax?],
+    unexpectedNodeCondition: (S) -> Bool,
+    correctNodes: [S?],
     message: (_ misplacedTokens: [TokenSyntax]) -> some DiagnosticMessage,
     moveFixIt: (_ misplacedTokens: [TokenSyntax]) -> FixItMessage,
     removeRedundantFixIt: (_ misplacedTokens: [TokenSyntax]) -> FixItMessage? = { _ in nil }
   ) {
     guard let incorrectContainer = unexpected else { return }
-    let misplacedTokens = incorrectContainer.presentTokens(satisfying: unexpectedTokenCondition)
-    if misplacedTokens.isEmpty { return }
+    let misplacedNodes = incorrectContainer.compactMap { $0.as(S.self) }.filter(unexpectedNodeCondition)
+    let misplacedTokens = misplacedNodes.flatMap {
+      Array($0.tokens(viewMode: .sourceAccurate))
+    }
```

Comment:
Couldn’t this just be `[Syntax]` as well? `FixIt.MultiNodeChange.makeMissing` supports nodes as well as tokens and I assume that `message`, `moveFixIt`, and `removeRedundantFixIt` could be changed to take `[Syntax]`. And `misplacedTokens.only` could become `misplacedTokens.only.as(TokenSyntax.self)`.

Reply:
Ah, that’s unfortunate. I was hoping that the changes to the `message` etc. closures would be minimal. I still think it would be good to migrate them to `Syntax` but that doesn’t need to be part of this PR. Could you file an issue for that so we can do the migration later?